### DISCUSSION TITLE
How to use the Python module

### DISCUSSION THREAD

**Discussion by ivo-s** [fontforge/fontforge#4566] [Q&A] [answered]
Hello, in 2019 I was able to somehow import the `fontforge` module into Python, as I have some scripts that worked at the time. I found some older discussions mentioning the corresponding `./configure` parameters. Now I wanted to re-use a certain script again and I found out that I cannot find the FontForge Python module anywhere. I also can't find any executable `ffpython`. I am on Linux Mint, also using Anaconda, and I built FontForge from the git repo. I only managed to run the script by switching to a system-provided Python 2 that somehow contained the fontforge library. It would be really helpful if someone clarified the status of the Python module and how to get it working.

**Comment by skef** [comment, marked as answer]
I think the last time a similar topic came up it appeared that the Linux Mint package was very old, and was only built for python2 for that reason. You should be able to compile it yourself, though. [This](https://github.com/fontforge/fontforge/blob/master/INSTALL.md) has some basic info and [this](https://github.com/fontforge/fontforge/wiki/CMake-guide-for-FontForge) breaks out the specific CMake parameters. However, I believe the python modules are built by default for the version corresponding to "`python`" (or whatever you set the `PYTHON` environment variable to when building) in the path.

**Reply by ivo-s** [reply]
Thank you for the links, I found out that the library is built as `fontforge.so`, which I didn't know could be a Python module. However, it is installed in `/usr/local/lib/python3/dist-packages/`. This particular path is not in Python's `sys.path`. I tried moving all the files I could find relevant, fontforge, psMat, pyhook, .so and .la, into a proper folder. Then, I got a new error in response to `import fontforge`:
`ImportError: dynamic module does not define module export function (PyInit_fontforge)`
I am probably missing more files, but I don't know which ones go where.

I also tried the cmake options specifying the Python version and the options `-DPYTHON_INCLUDE_DIR, -DPYTHON_LIBRARY` accordig to the Wiki, and running install, but it didn't seem to help. To be sure I am not doing anything non-standard, I deactivated Anaconda completely and installed the package `python-is-python3`.

**Reply by ivo-s** [reply]
For future reference, it seems that the standard building and installation procedure described in `INSTALL.md` should do the trick. For users with Anaconda, I recommend turning it off by commenting out the respective section in `~/.bashrc` marked by `>>> conda initialize >>>`. Advanced users may benefit from the CMake parameters documentation linked above.

**Comment by skef** [comment]
Can you post the output of `nm fontforge.so`?

**Reply by ivo-s** [reply]
I deleted the libraries I manually copied including `fontforge.so`, so now I cleared the whole build directory, rebuilt and reinstalled...and it started working! (coincidentally, the function `PyInit_fontforge` is now present, but the output of `nm` is different than I remember from before, perhaps I somehow copied an incorrect file there before). Much confused, I started tracking down what actually caused the change. It is possible that I haven't done the building and installing from scratch after completely turning off Anaconda and installing `python-is-python3`. It seems the package is not needed after all, while Anaconda does need to be turned off in order to avoid library overshadowing and specifying paths manually.

The library is still installed in the same directory (outside of sys.path), but this time, Python finds it, and if I delete it, Python complains.

All in all, I can't seem to reproduce the problem, now everything works stubbornly :smiley: and only using the simple build commands from the Install.md. Thank you for that Wiki link though, I hadn't noticed that part of the documentation. Unfortunately, I can't make any useful conclusion from all this, because it seems that I just had some unknown individual issue that is now resolved. Thank you for the help!
